northstar/src/runtime/minijail.rs

```diff
@@ -509,84 +521,93 @@ impl Debug {
     }
 }
 
-/// Pipe with async tokio impls
-#[derive(Debug)]
-struct AsyncPipe {
-    non_blocking: AsyncFd<std::fs::File>,
-    blocking: RawFd,
+/// Creates a pipe with the read endpoint implementing tokio::io::AsyncRead
+async fn create_pipe() -> Result<(PipeRead, PipeWrite), Error> {
+    let (readfd, writefd) =
+        pipe2(OFlag::O_NONBLOCK).map_err(|e| Error::Os("Failed to create pipe".to_string(), e))?;
+    let async_read =
+        AsyncFd::new(PipeFd(readfd)).map_err(|e| Error::Io("Async fd".to_string(), e))?;
+    Ok((PipeRead(async_read), PipeWrite(PipeFd(writefd))))
 }
 
-impl AsyncPipe {
-    fn new() -> Result<AsyncPipe, Error> {
-        let (non_blocking, blocking) =
-            pipe().map_err(|e| Error::Os("Failed to create pipe".to_string(), e))?;
-
-        let mut flags =
-            OFlag::from_bits(fcntl(non_blocking, fcntl::FcntlArg::F_GETFL).unwrap()).unwrap();
-        flags.set(OFlag::O_NONBLOCK, true);
-        fcntl(non_blocking, fcntl::FcntlArg::F_SETFL(flags)).expect("Failed to configure pipe fd");
-        let non_blocking = unsafe { std::fs::File::from_raw_fd(non_blocking) };
-        let non_blocking =
-            AsyncFd::new(non_blocking).map_err(|e| Error::Io("Async fd".to_string(), e))?;
-
-        Ok(AsyncPipe {
-            non_blocking,
-            blocking,
-        })
-    }
+#[derive(Debug)]
+struct PipeFd(RawFd);
```

Comment:
From the semantic point of view this is a `RawFd` that is closed on drop. It doesn't have any relation to pipes at all.

Can we do the close explicitly and remove `PipeFd`? This is just the case in `shutdown`.

Reply:
It is a new type rather than an alias, it holds one of the fds of the pipe, therefore the name. I wanted some type to manage the lifetime of the fd to be sure the pipe would always be closed.

We can call it differently if you want (maybe `FdGuard` or something) but manually closing the fds feels like it can potentially cause similar problems that the ones we had :(